lib/teleterm/daemon/gateway_cert_reissuer_test.go

```diff
@@ -0,0 +1,292 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package daemon
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	api "github.com/gravitational/teleport/lib/teleterm/api/protogen/golang/v1"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+	"github.com/gravitational/teleport/lib/teleterm/clusters"
+	"github.com/gravitational/teleport/lib/teleterm/gateway"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+var log = logrus.WithField(trace.Component, "reissuer")
+
+func TestReissueCert_CallsDBCertReissuerOnceIfItReturnsSuccessfully(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	tshdEventsClient := &mockTSHDEventsClient{
+		callCounts: make(map[string]int),
+	}
+	reissuer := &GatewayCertReissuer{
+		Log:              log,
+		TSHDEventsClient: tshdEventsClient,
+	}
+	gateway := mustCreateGateway(ctx, t)
+	dbCertReissuer := &mockDBCertReissuer{}
+
+	err := reissuer.ReissueCert(ctx, gateway, dbCertReissuer)
+	require.NoError(t, err)
+
+	require.Equal(t, 1, dbCertReissuer.callCount,
+		"Expected DBCertReissuer to have been called exactly one time")
+	require.Equal(t, 0, tshdEventsClient.callCounts["Relogin"],
+		"Expected TSHDEventsClient.Relogin to have been called exactly zero times")
+	require.Equal(t, 0, tshdEventsClient.callCounts["SendNotification"],
+		"Expected TSHDEventsClient.SendNotification to have been called exactly zero times")
+}
```

Comment:
I didn't opt for table tests here. I feel like the differences between test cases here are so specific that if I had to write a single test body that accounts for all seven cases, it'd be much harder to understand what's happening in each test case vs being able to follow the explicit setup of each test case.

Reply:
Well, this version of tests is so much better, thank you very much! And it disproves what I said. I thought that with table tests it'd be much harder to represent all those little differences between tests but it makes them even easier to notice.

I definitely underestimated how much heavy lifting the default values are going to do here.

I'm going to use it with a few modifications. I'll call `require.True(t, reissuer.reloginMu.TryLock())` rather than using `t.Fatalf` there because `t.Fatalf` shouldn't be used in subtests AFAIK. And I'll add some basic error messages to the `require`s related to call count so that it's easier to see which check failed.

Reply:
Actually, FailNow there is fine I think, I just tested it and didn't get any warnings. It is called from a goroutine running the test.

> FailNow must be called from the goroutine running the test or benchmark function, not from other goroutines created during the test. Calling FailNow does not stop those other goroutines.

https://pkg.go.dev/testing#T.FailNow